File: source/apps/ios-app/oid4vc/oid4vc/Protocol/Iso18013_5/Ble/BleConstants.swift
/*
 * Copyright 2025 - 2026 OmniOne.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Foundation
import CoreBluetooth

/// Constants for ISO 18013-5 BLE communication.
/// Updated to match the specific implementation used in the Android project.
public struct BleConstants {
    /// Service UUID: 0000fff0-0000-1000-8000-00805f9b34fb
    public static let SERVICE_UUID = CBUUID(string: "0000fff0-0000-1000-8000-00805f9b34fb")
    
    /// State characteristic UUID: 00000001-a123-48ce-896b-4c76973373e6
    public static let CHARACTERISTIC_STATE_UUID = CBUUID(string: "00000001-a123-48ce-896b-4c76973373e6")
    
    /// Client2Server characteristic UUID: 00000002-a123-48ce-896b-4c76973373e6
    public static let CHARACTERISTIC_CLIENT_2_SERVER_UUID = CBUUID(string: "00000002-a123-48ce-896b-4c76973373e6")
    
    /// Server2Client characteristic UUID: 00000003-a123-48ce-896b-4c76973373e6
    public static let CHARACTERISTIC_SERVER_2_CLIENT_UUID = CBUUID(string: "00000003-a123-48ce-896b-4c76973373e6")
    
    /// Ident characteristic UUID: 00000004-a123-48ce-896b-4c76973373e6
    public static let CHARACTERISTIC_IDENT_UUID = CBUUID(string: "00000004-a123-48ce-896b-4c76973373e6")
    
    /// State: Start (Reader ready to receive request)
    public static let STATE_START: UInt8 = 0x01
    
    /// State: End (Session terminated)
    public static let STATE_END: UInt8 = 0x02
    
    /// Client Characteristic Configuration Descriptor (standard)
    public static let DESC_CLIENT_CHAR_CONFIG = CBUUID(string: "2902")
}

File: source/sdks/poc-mso-mdoc-sdk-aos/src/main/java/org/omnione/did/sdk/mdoc/proximity/holder/ble/BleConstants.java
package org.omnione.did.sdk.mdoc.proximity.holder.ble;

import java.nio.ByteBuffer;
import java.util.UUID;

public class BleConstants {
    public static final UUID SERVICE_UUID = UUID.fromString("0000fff0-0000-1000-8000-00805f9b34fb");

    // Holder = GATT server (Reader = central client)
    public static final UUID CHARACTERISTIC_STATE_UUID = UUID.fromString("00000001-a123-48ce-896b-4c76973373e6");
    public static final UUID CHARACTERISTIC_CLIENT_2_SERVER_UUID = UUID.fromString("00000002-a123-48ce-896b-4c76973373e6");
    public static final UUID CHARACTERISTIC_SERVER_2_CLIENT_UUID = UUID.fromString("00000003-a123-48ce-896b-4c76973373e6");
    public static final UUID CHARACTERISTIC_IDENT_UUID = UUID.fromString("00000004-a123-48ce-896b-4c76973373e6");

    // Holder = central client (Reader = GATT server)
    public static final UUID READER_CHARACTERISTIC_STATE_UUID = UUID.fromString("00000005-a123-48ce-896b-4c76973373e6");
    public static final UUID READER_CHARACTERISTIC_CLIENT_2_SERVER_UUID = UUID.fromString("00000006-a123-48ce-896b-4c76973373e6");
    public static final UUID READER_CHARACTERISTIC_SERVER_2_CLIENT_UUID = UUID.fromString("00000007-a123-48ce-896b-4c76973373e6");
    public static final UUID READER_CHARACTERISTIC_IDENT_UUID = UUID.fromString("00000008-a123-48ce-896b-4c76973373e6");

    // Descriptor UUID for Notifications (BLE 표준)
    public static final UUID DESC_CLIENT_CHAR_CONFIG = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb");

    // State values
    public static final byte STATE_START = 0x01;
    public static final byte STATE_END = 0x02;

    /**
     * Generates a random 16-byte UUID for BLE communication.
     * @return 16-byte array representing the UUID.
     */
    public static byte[] generateRandomBleUuid() {
        UUID uuid = UUID.randomUUID();
        ByteBuffer buffer = ByteBuffer.allocate(16);
        buffer.putLong(uuid.getMostSignificantBits());
        buffer.putLong(uuid.getLeastSignificantBits());
        return buffer.array();
    }
}
